# CitiBike Data Analysis 
## Shakhawat Hassan

In [42]:
#importing libraries
from timeit import default_timer as timer
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
%matplotlib inline

In [43]:
df = pd.read_csv ("march20_tripdata.csv")

### Data Exploration

In [44]:
df.shape #prints number of rows and columns in DataFrame

(17719, 15)

In [45]:
df.head() #prints first n rows of the DataFrame

tripduration                 starttime                  stoptime  \
0           389  2020-03-01 00:14:07.0000  2020-03-01 00:20:36.0880   
1           242  2020-03-01 00:48:06.4390  2020-03-01 00:52:08.9870   
2           124  2020-03-01 01:08:46.4880  2020-03-01 01:10:51.2780   
3           104  2020-03-01 01:22:35.4790  2020-03-01 01:24:20.2470   
4           228  2020-03-01 01:39:20.2680  2020-03-01 01:43:09.1380   

   start station id start station name  start station latitude  \
0              3202       Newport PATH               40.727224   
1              3185          City Hall               40.717733   
2              3272       Jersey & 3rd               40.723332   
3              3202       Newport PATH               40.727224   
4              3194    McGinley Square               40.725340   

   start station longitude  end station id   end station name  \
0               -74.033759            3203      Hamilton Park   
1               -74.043845            3205  JC Medical Center   
2               -74.045953            3278   Monmouth and 6th   
3               -74.033759            3638      Washington St   
4               -74.067622            3280        Astor Place   

   end station latitude  end station longitude  bikeid    usertype  \
0             40.727596             -74.044247   42381  Subscriber   
1             40.716540             -74.049638   42155  Subscriber   
2             40.725685             -74.048790   42376  Subscriber   
3             40.724294             -74.035483   42350  Subscriber   
4             40.719282             -74.071262   42235  Subscriber   

   birth year  gender  
0        1992       1  
1        1991       1  
2        1987       0  
3        1993       1  
4        1988       1

In [46]:
df.tail() #prints last n rows of the DataFrame

tripduration                 starttime                  stoptime  \
17714           296  2020-03-31 22:50:39.0410  2020-03-31 22:55:35.5980   
17715           147  2020-03-31 23:09:15.9690  2020-03-31 23:11:43.5310   
17716           364  2020-03-31 23:17:02.0240  2020-03-31 23:23:06.8200   
17717          1217  2020-03-31 23:36:10.6430  2020-03-31 23:56:27.8120   
17718           846  2020-03-31 23:37:21.1840  2020-03-31 23:51:28.1220   

       start station id start station name  start station latitude  \
17714              3191           Union St               40.718211   
17715              3640     Journal Square               40.733670   
17716              3195            Sip Ave               40.730897   
17717              3267       Morris Canal               40.712419   
17718              3195            Sip Ave               40.730897   

       start station longitude  end station id  end station name  \
17714               -74.083639            3694    Jackson Square   
17715               -74.062500            3206           Hilltop   
17716               -74.063913            3679        Bergen Ave   
17717               -74.038526            3276  Marin Light Rail   
17718               -74.063913            3193      Lincoln Park   

       end station latitude  end station longitude  bikeid    usertype  \
17714             40.711130             -74.078900   42145  Subscriber   
17715             40.731169             -74.057574   42110  Subscriber   
17716             40.722104             -74.071455   42550  Subscriber   
17717             40.714584             -74.042817   42115  Subscriber   
17718             40.724605             -74.078406   42362  Subscriber   

       birth year  gender  
17714        1994       1  
17715        1991       1  
17716        1986       1  
17717        1986       1  
17718        1965       1

In [47]:
df.info() #data type and memory information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17719 entries, 0 to 17718
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   tripduration             17719 non-null  int64  
 1   starttime                17719 non-null  object 
 2   stoptime                 17719 non-null  object 
 3   start station id         17719 non-null  int64  
 4   start station name       17719 non-null  object 
 5   start station latitude   17719 non-null  float64
 6   start station longitude  17719 non-null  float64
 7   end station id           17719 non-null  int64  
 8   end station name         17719 non-null  object 
 9   end station latitude     17719 non-null  float64
 10  end station longitude    17719 non-null  float64
 11  bikeid                   17719 non-null  int64  
 12  usertype                 17719 non-null  object 
 13  birth year               17719 non-null  int64  
 14  gender                

In [48]:
df.describe() #summary statistics for numerical columns

tripduration  start station id  start station latitude  \
count  1.771900e+04      17719.000000            17719.000000   
mean   1.384512e+03       3307.352051               40.722406   
std    2.611010e+04        183.438917                0.007006   
min    6.100000e+01       3184.000000               40.711130   
25%    2.520000e+02       3195.000000               40.717733   
50%    4.200000e+02       3211.000000               40.721525   
75%    8.280000e+02       3276.000000               40.727224   
max    1.930668e+06       3792.000000               40.748716   

       start station longitude  end station id  end station latitude  \
count             17719.000000    17719.000000          17719.000000   
mean                -74.047013     3305.887748             40.722134   
std                   0.011588      183.892897              0.006971   
min                 -74.083639     3184.000000             40.709651   
25%                 -74.050990     3195.000000             40.717733   
50%                 -74.044247     3211.000000             40.721124   
75%                 -74.038051     3276.000000             40.726012   
max                 -74.032108     3792.000000             40.814394   

       end station longitude        bikeid    birth year        gender  
count           17719.000000  17719.000000  17719.000000  17719.000000  
mean              -74.046631  42103.189740   1980.739489      1.152379  
std                 0.011576   1999.092587     10.768536      0.546236  
min               -74.083639  14578.000000   1939.000000      0.000000  
25%               -74.050656  42213.000000   1972.000000      1.000000  
50%               -74.043845  42348.000000   1983.000000      1.000000  
75%               -74.038051  42477.000000   1989.000000      1.000000  
max               -73.953247  42636.000000   2002.000000      2.000000

In [49]:
df.mean() #Returns the mean of all columns

tripduration                1384.512444
start station id            3307.352051
start station latitude        40.722406
start station longitude      -74.047013
end station id              3305.887748
end station latitude          40.722134
end station longitude        -74.046631
bikeid                     42103.189740
birth year                  1980.739489
gender                         1.152379
dtype: float64

In [50]:
df.corr() #Returns the correlation between columns in a DataFrame

tripduration  start station id  \
tripduration                 1.000000          0.019928   
start station id             0.019928          1.000000   
start station latitude      -0.005883         -0.099169   
start station longitude     -0.008527          0.111445   
end station id               0.005356          0.058657   
end station latitude        -0.009192          0.027498   
end station longitude       -0.014715          0.080373   
bikeid                      -0.005801          0.013431   
birth year                  -0.021978         -0.009692   
gender                      -0.044914         -0.032407   

                         start station latitude  start station longitude  \
tripduration                          -0.005883                -0.008527   
start station id                      -0.099169                 0.111445   
start station latitude                 1.000000                -0.218648   
start station longitude               -0.218648                 1.000000   
end station id                         0.026595                 0.059414   
end station latitude                   0.422789                -0.200341   
end station longitude                 -0.228921                 0.567427   
bikeid                                -0.051511                 0.023460   
birth year                             0.032483                -0.001658   
gender                                 0.015078                 0.053736   

                         end station id  end station latitude  \
tripduration                   0.005356             -0.009192   
start station id               0.058657              0.027498   
start station latitude         0.026595              0.422789   
start station longitude        0.059414             -0.200341   
end station id                 1.000000             -0.085725   
end station latitude          -0.085725              1.000000   
end station longitude          0.127766             -0.178206   
bikeid                        -0.001557             -0.033481   
birth year                    -0.014956              0.027031   
gender                        -0.048203              0.018220   

                         end station longitude    bikeid  birth year    gender  
tripduration                         -0.014715 -0.005801   -0.021978 -0.044914  
start station id                      0.080373  0.013431   -0.009692 -0.032407  
start station latitude               -0.228921 -0.051511    0.032483  0.015078  
start station longitude               0.567427  0.023460   -0.001658  0.053736  
end station id                        0.127766 -0.001557   -0.014956 -0.048203  
end station latitude                 -0.178206 -0.033481    0.027031  0.018220  
end station longitude                 1.000000  0.032270    0.004544  0.040392  
bikeid                                0.032270  1.000000    0.000787  0.008902  
birth year                            0.004544  0.000787    1.000000  0.234028  
gender                                0.040392  0.008902    0.234028  1.000000

In [51]:
df.count() #Returns the number of non-null values in each DataFrame column

tripduration               17719
starttime                  17719
stoptime                   17719
start station id           17719
start station name         17719
start station latitude     17719
start station longitude    17719
end station id             17719
end station name           17719
end station latitude       17719
end station longitude      17719
bikeid                     17719
usertype                   17719
birth year                 17719
gender                     17719
dtype: int64

In [52]:
df.max() #Returns the highest value in each column

tripduration                                1930668
starttime                  2020-03-31 23:37:21.1840
stoptime                   2020-04-16 20:34:46.7660
start station id                               3792
start station name                          York St
start station latitude                      40.7487
start station longitude                    -74.0321
end station id                                 3792
end station name                            York St
end station latitude                        40.8144
end station longitude                      -73.9532
bikeid                                        42636
usertype                                 Subscriber
birth year                                     2002
gender                                            2
dtype: object

In [53]:
df.min() #Returns the lowest value in each column

tripduration                                     61
starttime                  2020-03-01 00:14:07.0000
stoptime                   2020-03-01 00:20:36.0880
start station id                               3184
start station name                5 Corners Library
start station latitude                      40.7111
start station longitude                    -74.0836
end station id                                 3184
end station name                  5 Corners Library
end station latitude                        40.7097
end station longitude                      -74.0836
bikeid                                        14578
usertype                                   Customer
birth year                                     1939
gender                                            0
dtype: object

In [54]:
df.median() #Returns the median in each column

tripduration                 420.000000
start station id            3211.000000
start station latitude        40.721525
start station longitude      -74.044247
end station id              3211.000000
end station latitude          40.721124
end station longitude        -74.043845
bikeid                     42348.000000
birth year                  1983.000000
gender                         1.000000
dtype: float64

In [55]:
df.std()

tripduration               26110.100952
start station id             183.438917
start station latitude         0.007006
start station longitude        0.011588
end station id               183.892897
end station latitude           0.006971
end station longitude          0.011576
bikeid                      1999.092587
birth year                    10.768536
gender                         0.546236
dtype: float64

### Data Cleaning

In [56]:
df.isnull() #checking for null values

tripduration  starttime  stoptime  start station id  \
0             False      False     False             False   
1             False      False     False             False   
2             False      False     False             False   
3             False      False     False             False   
4             False      False     False             False   
...             ...        ...       ...               ...   
17714         False      False     False             False   
17715         False      False     False             False   
17716         False      False     False             False   
17717         False      False     False             False   
17718         False      False     False             False   

       start station name  start station latitude  start station longitude  \
0                   False                   False                    False   
1                   False                   False                    False   
2                   False                   False                    False   
3                   False                   False                    False   
4                   False                   False                    False   
...                   ...                     ...                      ...   
17714               False                   False                    False   
17715               False                   False                    False   
17716               False                   False                    False   
17717               False                   False                    False   
17718               False                   False                    False   

       end station id  end station name  end station latitude  \
0               False             False                 False   
1               False             False                 False   
2               False             False                 False   
3               False             False                 False   
4               False             False                 False   
...               ...               ...                   ...   
17714           False             False                 False   
17715           False             False                 False   
17716           False             False                 False   
17717           False             False                 False   
17718           False             False                 False   

       end station longitude  bikeid  usertype  birth year  gender  
0                      False   False     False       False   False  
1                      False   False     False       False   False  
2                      False   False     False       False   False  
3                      False   False     False       False   False  
4                      False   False     False       False   False  
...                      ...     ...       ...         ...     ...  
17714                  False   False     False       False   False  
17715                  False   False     False       False   False  
17716                  False   False     False       False   False  
17717                  False   False     False       False   False  
17718                  False   False     False       False   False  

[17719 rows x 15 columns]

In [57]:
df.notnull() #checking for not-null values

tripduration  starttime  stoptime  start station id  \
0              True       True      True              True   
1              True       True      True              True   
2              True       True      True              True   
3              True       True      True              True   
4              True       True      True              True   
...             ...        ...       ...               ...   
17714          True       True      True              True   
17715          True       True      True              True   
17716          True       True      True              True   
17717          True       True      True              True   
17718          True       True      True              True   

       start station name  start station latitude  start station longitude  \
0                    True                    True                     True   
1                    True                    True                     True   
2                    True                    True                     True   
3                    True                    True                     True   
4                    True                    True                     True   
...                   ...                     ...                      ...   
17714                True                    True                     True   
17715                True                    True                     True   
17716                True                    True                     True   
17717                True                    True                     True   
17718                True                    True                     True   

       end station id  end station name  end station latitude  \
0                True              True                  True   
1                True              True                  True   
2                True              True                  True   
3                True              True                  True   
4                True              True                  True   
...               ...               ...                   ...   
17714            True              True                  True   
17715            True              True                  True   
17716            True              True                  True   
17717            True              True                  True   
17718            True              True                  True   

       end station longitude  bikeid  usertype  birth year  gender  
0                       True    True      True        True    True  
1                       True    True      True        True    True  
2                       True    True      True        True    True  
3                       True    True      True        True    True  
4                       True    True      True        True    True  
...                      ...     ...       ...         ...     ...  
17714                   True    True      True        True    True  
17715                   True    True      True        True    True  
17716                   True    True      True        True    True  
17717                   True    True      True        True    True  
17718                   True    True      True        True    True  

[17719 rows x 15 columns]

In [58]:
df.dropna() #drop all rows that contain null values

tripduration                 starttime                  stoptime  \
0               389  2020-03-01 00:14:07.0000  2020-03-01 00:20:36.0880   
1               242  2020-03-01 00:48:06.4390  2020-03-01 00:52:08.9870   
2               124  2020-03-01 01:08:46.4880  2020-03-01 01:10:51.2780   
3               104  2020-03-01 01:22:35.4790  2020-03-01 01:24:20.2470   
4               228  2020-03-01 01:39:20.2680  2020-03-01 01:43:09.1380   
...             ...                       ...                       ...   
17714           296  2020-03-31 22:50:39.0410  2020-03-31 22:55:35.5980   
17715           147  2020-03-31 23:09:15.9690  2020-03-31 23:11:43.5310   
17716           364  2020-03-31 23:17:02.0240  2020-03-31 23:23:06.8200   
17717          1217  2020-03-31 23:36:10.6430  2020-03-31 23:56:27.8120   
17718           846  2020-03-31 23:37:21.1840  2020-03-31 23:51:28.1220   

       start station id start station name  start station latitude  \
0                  3202       Newport PATH               40.727224   
1                  3185          City Hall               40.717733   
2                  3272       Jersey & 3rd               40.723332   
3                  3202       Newport PATH               40.727224   
4                  3194    McGinley Square               40.725340   
...                 ...                ...                     ...   
17714              3191           Union St               40.718211   
17715              3640     Journal Square               40.733670   
17716              3195            Sip Ave               40.730897   
17717              3267       Morris Canal               40.712419   
17718              3195            Sip Ave               40.730897   

       start station longitude  end station id   end station name  \
0                   -74.033759            3203      Hamilton Park   
1                   -74.043845            3205  JC Medical Center   
2                   -74.045953            3278   Monmouth and 6th   
3                   -74.033759            3638      Washington St   
4                   -74.067622            3280        Astor Place   
...                        ...             ...                ...   
17714               -74.083639            3694     Jackson Square   
17715               -74.062500            3206            Hilltop   
17716               -74.063913            3679         Bergen Ave   
17717               -74.038526            3276   Marin Light Rail   
17718               -74.063913            3193       Lincoln Park   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.727596             -74.044247   42381  Subscriber   
1                 40.716540             -74.049638   42155  Subscriber   
2                 40.725685             -74.048790   42376  Subscriber   
3                 40.724294             -74.035483   42350  Subscriber   
4                 40.719282             -74.071262   42235  Subscriber   
...                     ...                    ...     ...         ...   
17714             40.711130             -74.078900   42145  Subscriber   
17715             40.731169             -74.057574   42110  Subscriber   
17716             40.722104             -74.071455   42550  Subscriber   
17717             40.714584             -74.042817   42115  Subscriber   
17718             40.724605             -74.078406   42362  Subscriber   

       birth year  gender  
0            1992       1  
1            1991       1  
2            1987       0  
3            1993       1  
4            1988       1  
...           ...     ...  
17714        1994       1  
17715        1991       1  
17716        1986       1  
17717        1986       1  
17718        1965       1  

[17719 rows x 15 columns]

In [59]:
df.dropna(axis=1) #drop all columns that contain null values

tripduration                 starttime                  stoptime  \
0               389  2020-03-01 00:14:07.0000  2020-03-01 00:20:36.0880   
1               242  2020-03-01 00:48:06.4390  2020-03-01 00:52:08.9870   
2               124  2020-03-01 01:08:46.4880  2020-03-01 01:10:51.2780   
3               104  2020-03-01 01:22:35.4790  2020-03-01 01:24:20.2470   
4               228  2020-03-01 01:39:20.2680  2020-03-01 01:43:09.1380   
...             ...                       ...                       ...   
17714           296  2020-03-31 22:50:39.0410  2020-03-31 22:55:35.5980   
17715           147  2020-03-31 23:09:15.9690  2020-03-31 23:11:43.5310   
17716           364  2020-03-31 23:17:02.0240  2020-03-31 23:23:06.8200   
17717          1217  2020-03-31 23:36:10.6430  2020-03-31 23:56:27.8120   
17718           846  2020-03-31 23:37:21.1840  2020-03-31 23:51:28.1220   

       start station id start station name  start station latitude  \
0                  3202       Newport PATH               40.727224   
1                  3185          City Hall               40.717733   
2                  3272       Jersey & 3rd               40.723332   
3                  3202       Newport PATH               40.727224   
4                  3194    McGinley Square               40.725340   
...                 ...                ...                     ...   
17714              3191           Union St               40.718211   
17715              3640     Journal Square               40.733670   
17716              3195            Sip Ave               40.730897   
17717              3267       Morris Canal               40.712419   
17718              3195            Sip Ave               40.730897   

       start station longitude  end station id   end station name  \
0                   -74.033759            3203      Hamilton Park   
1                   -74.043845            3205  JC Medical Center   
2                   -74.045953            3278   Monmouth and 6th   
3                   -74.033759            3638      Washington St   
4                   -74.067622            3280        Astor Place   
...                        ...             ...                ...   
17714               -74.083639            3694     Jackson Square   
17715               -74.062500            3206            Hilltop   
17716               -74.063913            3679         Bergen Ave   
17717               -74.038526            3276   Marin Light Rail   
17718               -74.063913            3193       Lincoln Park   

       end station latitude  end station longitude  bikeid    usertype  \
0                 40.727596             -74.044247   42381  Subscriber   
1                 40.716540             -74.049638   42155  Subscriber   
2                 40.725685             -74.048790   42376  Subscriber   
3                 40.724294             -74.035483   42350  Subscriber   
4                 40.719282             -74.071262   42235  Subscriber   
...                     ...                    ...     ...         ...   
17714             40.711130             -74.078900   42145  Subscriber   
17715             40.731169             -74.057574   42110  Subscriber   
17716             40.722104             -74.071455   42550  Subscriber   
17717             40.714584             -74.042817   42115  Subscriber   
17718             40.724605             -74.078406   42362  Subscriber   

       birth year  gender  
0            1992       1  
1            1991       1  
2            1987       0  
3            1993       1  
4            1988       1  
...           ...     ...  
17714        1994       1  
17715        1991       1  
17716        1986       1  
17717        1986       1  
17718        1965       1  

[17719 rows x 15 columns]

In [62]:
df.isnull().sum() #null checking

tripduration               0
starttime                  0
stoptime                   0
start station id           0
start station name         0
start station latitude     0
start station longitude    0
end station id             0
end station name           0
end station latitude       0
end station longitude      0
bikeid                     0
usertype                   0
birth year                 0
gender                     0
dtype: int64

In [64]:
df.duplicated().sum() #looking for duplicates

0

In [67]:
df['bikeid'].count()

17719

#### There are 17,719 users who rented Citi Bikes in March 2020

In [70]:
df['tripduration'].nlargest()

15899    1930668
16170    1796894
13847    1631164
17466    1020814
16245     784008
Name: tripduration, dtype: int64